ateam_geometry/src/segment.cpp

```diff
@@ -0,0 +1,232 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+#include <math.h>
+#include <Eigen/Dense>
+
+#include <optional>
+#include <vector>
+
+#include "ateam_geometry/segment.hpp"
+#include "ateam_geometry/utilities.hpp"
+#include "ateam_common/equality_utilities.hpp"
+
+namespace ateam_geometry
+{
+LineSegment::LineSegment(const Eigen::Vector2d & p1, const Eigen::Vector2d & p2)
+{
+  this->p1 = p1;
+  this->p2 = p2;
+}
+
+LineSegment::LineSegment(
+  const Eigen::Vector2d & start, const double & length,
+  const double & angle)
+{
+  p1 = start;
+  Eigen::Vector2d endpoint;
+  // Essentially we convert a given polar coordinate/vector
+  // (length and angle) to an x, y plane
+  endpoint.x() = start.x() + (length * cos(angle));
+  endpoint.y() = start.y() + (length * sin(angle));
+  p2 = endpoint;
+}
+
+double LineSegment::get_length()
+{
+  // Length is Euclidian, we can change later if desired
+  return (p2 - p1).norm();
+}
+
+Eigen::Vector2d LineSegment::get_midpoint()
+{
+  return p1 + 0.5 * (p2 - p1);
+}
+
+std::vector<Eigen::Vector2d> LineSegment::get_equally_spaced_points(const int & num_points)
+{
+  /*
+      Get equally spaced points along the line segement this is called from.
+      If we only get one point, it will be the first endpoint.
+      */
+  std::vector<Eigen::Vector2d> points;
+  if (num_points == 1) {
+    points.push_back(this->p1);
+    return points;
+  }
+  Eigen::Vector2d spacing;
+  spacing.x() = (p2.x() - p1.x()) / (num_points - 1);
+  spacing.y() = (p2.y() - p1.y()) / (num_points - 1);
+  for (int i = 0; i < num_points; ++i) {
+    Eigen::Vector2d point = this->p1;
+    point = point + (spacing * i);
+    points.push_back(point);
+  }
+  return points;
+}
+
+bool is_point_on_segment(const Eigen::Vector2d & point, LineSegment & segment, double tolerance)
+{
+  /*
+      https://computergraphics.stackexchange.com/questions/2105/test-if-a-point-is-on-a-line-segment
+      https://lucidar.me/en/mathematics/check-if-a-point-belongs-on-a-line-segment/
+  */
+  double crossProduct = (point.y() - segment.p1.y()) * (segment.p2.x() - segment.p1.x()) -
+    (point.x() - segment.p1.x()) * (segment.p2.y() - segment.p1.y());
+
+  if (std::abs(crossProduct) > tolerance) {
```

Comment:
Unsure this tolerance should be used for both the dot prod and the cross product.